analyzers/src/SonarAnalyzer.CSharp/Rules/UnusedPrivateMember.cs

```diff
@@ -291,41 +292,79 @@ private static bool VisitDeclaringReferences(ISymbol symbol, ISafeSyntaxWalker v
                 syntaxReference.SyntaxTree.IsGenerated(CSharpGeneratedCodeRecognizer.Instance, compilation);
         }
 
+        private static CSharpRemovableSymbolWalker RetrieveRemovableSymbols(INamedTypeSymbol namedType, Compilation compilation)
+        {
+            var removableSymbolsCollector = new CSharpRemovableSymbolWalker(compilation.GetSemanticModel, namedType.DeclaredAccessibility);
+            if (!VisitDeclaringReferences(namedType, removableSymbolsCollector, compilation, includeGeneratedFile: false))
+            {
+                return null;
+            }
+
+            return removableSymbolsCollector;
+        }
+
+        private static void CopyRetrievedSymbols(CSharpRemovableSymbolWalker removableSymbolCollector,
+                                                 HashSet<ISymbol> privateSymbols,
+                                                 ConcurrentBag<ISymbol> internalSymbols,
+                                                 BidirectionalDictionary<ISymbol, SyntaxNode> fieldLikeSymbols)
+        {
+            privateSymbols.AddRange(removableSymbolCollector.PrivateSymbols);
+
+            // Keep the removable internal types for when the compilation ends
+            foreach (var internalSymbol in removableSymbolCollector.InternalSymbols.OfType<INamedTypeSymbol>())
```

Comment:
We are adding all internal symbols to this collection but we only need the ones that implement `INamedTypeSymbol`. Can't we add to the collection only the ones we actually need, and then remove this filtering?